runtime/Cpp/runtime/src/atn/SemanticContext.cpp

```diff
@@ -51,43 +51,79 @@ namespace {
     }
   }
 
-}
+  template <typename Comparer>
+  void insertSemanticContext(Ref<const SemanticContext> &&semanticContext,
+                             std::unordered_set<const SemanticContext*, SemanticContextHasher, SemanticContextComparer> &operandSet,
+                             std::vector<Ref<const SemanticContext>> &operandList,
+                             Ref<const SemanticContext::PrecedencePredicate> &precedencePredicate,
+                             Comparer comparer) {
+    if (semanticContext != nullptr) {
+      if (semanticContext->getContextType() == SemanticContextType::PRECEDENCE) {
+        if (precedencePredicate == nullptr || comparer(downCast<const SemanticContext::PrecedencePredicate*>(semanticContext.get())->precedence, precedencePredicate->precedence)) {
+          precedencePredicate = std::static_pointer_cast<const SemanticContext::PrecedencePredicate>(std::move(semanticContext));
+        }
+      } else {
+        auto [existing, inserted] = operandSet.insert(semanticContext.get());
+        if (inserted) {
+          operandList.push_back(std::move(semanticContext));
+        }
+      }
+    }
+  }
```

Comment:
You must have an `insertSemanticContext` function already, but you added another one. Why's that?

Reply:
This one accepts a rvalue reference `Ref<const SemanticContext>`, while the other takes a lvalue reference. This avoids an extra reference/unreference in `shared_ptr` under certain circumstances.

Reply:
I'm afraid of code duplication, but ok.